pkg/sqlite/sqlite.go

```diff
@@ -0,0 +1,163 @@
+package sqlite
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"os"
+	"time"
+
+	"github.com/adrg/xdg"
+	"github.com/docker/docker-credential-helpers/credentials"
+	"github.com/glebarez/sqlite"
+	"github.com/gptscript-ai/gptscript/pkg/config"
+	"gorm.io/gorm"
+	"gorm.io/gorm/logger"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/storage/value"
+)
+
+// uid is here to fulfill the value.Context interface for the transformer.
+// This is similar to authenticatedDataString from the k8s apiserver's storage interface
+// for etcd: https://github.com/kubernetes/kubernetes/blob/a42f4f61c2c46553bfe338eefe9e81818c7360b4/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L63
+type uid string
+
+func (u uid) AuthenticatedData() []byte {
+	return []byte(u)
+}
+
+var groupResource = schema.GroupResource{
+	Group:    "", // deliberately left empty
+	Resource: "credentials",
+}
+
+type Sqlite struct {
+	cfg         *config.CLIConfig
+	db          *gorm.DB
+	transformer value.Transformer
+}
+
+func NewSqlite(ctx context.Context) (Sqlite, error) {
+	// Passing the empty string here will first look for the config location in GPTSCRIPT_CONFIG_FILE.
+	// If that is not set, it will look at xdg.ConfigFile("gptscript/config.json"), the default location.
+	cfg, err := config.ReadCLIConfig("")
+	if err != nil {
+		return Sqlite{}, fmt.Errorf("error reading CLI config: %w", err)
+	}
+
+	var dbPath string
+	if os.Getenv("GPTSCRIPT_SQLITE_FILE") != "" {
+		dbPath = os.Getenv("GPTSCRIPT_SQLITE_FILE")
+	} else {
+		dbPath, err = xdg.ConfigFile("gptscript/credentials.db")
+		if err != nil {
+			return Sqlite{}, fmt.Errorf("failed to get credentials db path: %w", err)
+		}
+	}
+
+	db, err := gorm.Open(sqlite.Open(dbPath), &gorm.Config{
+		Logger: logger.New(log.New(os.Stdout, "\r\n", log.LstdFlags), logger.Config{
+			LogLevel:                  logger.Error,
+			IgnoreRecordNotFoundError: true,
+		}),
+	})
+	if err != nil {
+		return Sqlite{}, fmt.Errorf("failed to open database: %w", err)
+	}
+
+	if err := db.AutoMigrate(&GptscriptCredential{}); err != nil {
+		return Sqlite{}, fmt.Errorf("failed to auto migrate GptscriptCredential: %w", err)
+	}
+
+	s := Sqlite{
+		cfg: cfg,
+		db:  db,
+	}
+
+	encryptionConf, err := readEncryptionConfig(ctx)
+	if err != nil {
+		return Sqlite{}, fmt.Errorf("failed to read encryption config: %w", err)
+	} else if encryptionConf != nil {
+		transformer, exists := encryptionConf.Transformers[groupResource]
+		if !exists {
+			return Sqlite{}, fmt.Errorf("failed to find encryption transformer for %s", groupResource.String())
+		}
+		s.transformer = transformer
+	}
+
+	return s, nil
+}
+
+type GptscriptCredential struct {
+	ID        uint `gorm:"primary_key"`
+	CreatedAt time.Time
+	ServerURL string `gorm:"unique"`
```

Comment:
What is the serverURL in credential?

Reply:
This is so that we fit logically with how credentials work for the Docker credential helpers. Each Docker credential includes a server URL, username, and password. The server URL is basically the URL of the container store. In our case we use the server URL as a combination of the credential context and credential name. But the translation of that is handled in GPTScript itself, not here in the helper.

Reply:
thanks for the explanation!